numpy/core/src/multiarray/cblasfuncs.c

```diff
@@ -412,14 +413,38 @@ cblas_matrixproduct(int typenum, PyArrayObject *ap1, PyArrayObject *ap2,
                 goto fail;
             }
         }
+
+        /* check for memory overlap */
+        if (!(solve_may_share_memory(out, ap1, 1) == 0 &&
+              solve_may_share_memory(out, ap2, 1) == 0)) {
+            /* allocate temporary output array */
+            ret = (PyArrayObject *)PyArray_NewLikeArray(out, NPY_CORDER,
+                                                        NULL, 0);
+            if (ret == NULL) {
+                goto fail;
+            }
+
+            /* set copy-back */
+            Py_INCREF(out);
+            if (PyArray_SetUpdateIfCopyBase(ret, out) < 0) {
+                Py_DECREF(out);
+                goto fail;
+            }
+        }
+        else {
+            Py_INCREF(out);
+            ret = out;
+        }
         Py_INCREF(out);
```

Comment:
@charris: yes it's needed, since both variables are owned references.
@juliantaylor: agreed, would be best to change.